src/main/java/com/google/summit/SummitAST.kt

```diff
@@ -0,0 +1,199 @@
+        when {
+          isApexClassFile(path) -> CompilationType.CLASS
+          isApexTriggerFile(path) -> CompilationType.TRIGGER
+          else -> throw IllegalArgumentException("Unexpected file type")
+        },
+      charStream = CharStreams.fromPath(path),
+    )
+
+  /**
+   * Parses and translates the [string] as [type] and returns a [CompilationUnit] representing the
+   * AST if the operation was successful.
+   *
+   * If [type] is not provided, the compilation type is determined by the contents of the [string].
+   */
+  fun parseAndTranslate(string: String, type: CompilationType?) =
+    parseAndTranslate(
+      name = STRING_INPUT,
+      type = type,
+      charStream = CharStreams.fromString(string),
+    )
+
+  /**
+   * Parses and translates the [charStream] as [type] and returns a [CompilationUnit] representing
+   * the AST if the operation was successful.
+   *
+   * If [type] is not provided, the compilation type is determined by the contents of the
+   * [charStream].
+   */
+  internal fun parseAndTranslate(
+    name: String,
+    type: CompilationType?,
+    charStream: CharStream
+  ): CompilationUnit? {
+    // Apex is a case-insensitive language and the grammar is
+    // defined to operate on fully lower-cased inputs.
+    val lexer = ApexLexer(CaseInsensitiveInputStream(charStream))
+    val tokens = CommonTokenStream(lexer)
+    val parser = ApexParser(tokens)
+
+    val errorCounter = SyntaxErrorListener()
+    lexer.addErrorListener(errorCounter)
+    parser.addErrorListener(errorCounter)
+
+    // Do parse as complete compilation unit
+    val tree =
+      when (type ?: determineCompilationType(charStream)) {
+        CompilationType.CLASS -> parser.compilationUnit()
+        CompilationType.TRIGGER -> parser.triggerUnit()
+      }
+
+    if (errorCounter.numErrors > 0) {
+      logger.atWarning().log("Failed to parse %s", name)
+      return null // failure
+    }
+
+    try {
+      val translator = Translate(name, tokens)
+      val ast = translator.translate(tree)
+      return ast
+    } catch (e: Translate.TranslationException) {
+      logger.atWarning().withCause(e).log("Failed to translate %s", name)
+      return null // failure
+    }
+  }
+
+  /**
+   * Determines the [CompilationType] of [charStream].
+   *
+   * If `class` or `trigger` is found before the body of the top-level declaration, the
+   * corresponding [CompilationType] is returned. Otherwise, [CompilationType.CLASS] is returned by
+   * default.
+   *
+   * After execution, the input cursor of the [charStream] is reset to index 0.
+   */
+  private fun determineCompilationType(charStream: CharStream): CompilationType {
+    fun findType(): CompilationType? {
+      val lexer = ApexLexer(CaseInsensitiveInputStream(charStream))
+
+      // Discard tokens inside body of declaration
+      val tokens =
+        generateSequence { lexer.nextToken() }
+          .takeWhile { it.type != ApexLexer.LBRACE && it.type != ApexLexer.EOF }
```

Comment:
This does still read all tokens from the stream in the worst case (no `{` found). I think reading past the first token is an improvement in functionality, but is it worth the potential performance/memory impacts?

Reply:
That case isn't valid Apex code. In practice this will arise very infrequently. It's not worth optimizing performance.

LGTM.